lib/browser-pool/caching-pool.js

```diff
@@ -35,7 +36,13 @@ CachingPool.prototype.getBrowser = function(id) {
     }
 
     log('has cached browser %o', browser);
-    return browser.reset().thenResolve(browser);
+    return browser.reset()
+        .fail(function(e) {
+            var reject = q.reject.bind(null, e);
+            return this.freeBrowser(browser)
```

Comment:
`this.freeBrowser` will put it into the cache which is probably not what we want here. Better to call `underlyingPool.freeBrowser` here